### PR TITLE
Bluetooth: Mesh: Created API's for EMDS.

### PR DESCRIPTION
**Pull request by Balaklaka**
Create API's for the Emergency Data Storage. This storage should be have
its own storage area that can be quickly stored to at when necessary.
This way we can store data that is changing a lot without wearing out
the flash, and have a predefined time for storing the data. There should
be possible to determine how mush time is needed to store all data
configured to be stored in this area.

Signed-off-by: Ingar Kulbrandstad <ingar.kulbrandstad@nordicsemi.no>

### CODE REVIEW

File: include/emds/emds.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Comment (alxelax) [inline include/emds/emds.h:2]: 2022 came :). In other files too

File: include/emds/emds.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @struct emds_entry
+ *
+ * Information about entries used to store data in the emergency data storage.
+ */
+struct emds_entry {
```
Comment (alxelax) [inline include/emds/emds.h:31]: Should be any variable to keep the information it is static or dynamic entry?

File: include/emds/emds.h [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+ */
+#define EMDS_STATIC_ENTRY_DEFINE(_name, _id, _data, _len)                                          \
+	const STRUCT_SECTION_ITERABLE(emds_entry, emds_##_name) = {                                \
+		.id = _id,                                                                         \
+		.data = (uint8_t *)_data,                                                          \
+		.len = _len,                                                                       \
+	}
+
+/**
+ * @brief emds_init Initialize the emergency data storage.
```
Comment (alxelax) [inline include/emds/emds.h:70]: no need function name in description

**Comment by alxelax** [comment]
Shouldn't be added returned value to all API functions. It is not known in advance it will be required or not.

File: include/emds/emds.h [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+/**
+ * @brief Calculate the time needed to store the registered data.
+ *
+ * This will calculate how much time it takes to store all registered data, both
+ * dynamic and static data registered in the entries.
+ *
+ * @return Time in micro seconds that is needed to store all data.
+ */
+int emds_store_time_get(void);
```
Comment (alxelax) [inline include/emds/emds.h:147]: if time is in microseconds it is better to use a wider type. uint64_t probably?

File: include/emds/emds.h [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+/**
+ * @brief Calculate the size needed to store the registered data.
+ *
+ * This will calculate how much size it takes to store all registered data, both
+ * dynamic and static data registered in the entries.
+ *
+ * @return Byte size that is needed to store all data. 
+ */
+int emds_store_size_get(void);
```
Comment (alxelax) [inline include/emds/emds.h:157]: the same comment about the wide of the returned value.

File: subsys/emds/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+zephyr_sources(emds.c)
```
Comment (alxelax) [inline subsys/emds/CMakeLists.txt:7]: I guess it is better to use `zephyr_sources_ifdef` and dependency on the configuration option EMDS

**Comment by alxelax** [comment]
Why does it require separation on static and dynamic entries? In my opinion, is much better to have a unified entry and put then in the linked list. All entry descriptor operations will be the classic linked list algorithms. To have two types just complicates conception IMHO

**Review by alxelax — CHANGES_REQUESTED** [review]

File: subsys/emds/CMakeLists.txt (hunk 6 above)
Comment (Balaklaka) [reply subsys/emds/CMakeLists.txt:7]: My plan was to put that in the CMakeLists.txt file that takes the hole folder. So if EMDS is not added this file is not added.

File: include/emds/emds.h (hunk 4 above)
Comment (Balaklaka) [reply include/emds/emds.h:147]: It should be enough with a uint32_t, can not see that it is necessary with a uint64_t. This should not take more than 4000 seconds.

File: include/emds/emds.h (hunk 5 above)
Comment (Balaklaka) [reply include/emds/emds.h:157]: Can not see that this will be so large that it needs a uint64_t.

File: include/emds/emds.h (hunk 2 above)
Comment (Balaklaka) [reply include/emds/emds.h:31]: There is places that it does not fit to have this dynamic, like the RPL list. That does not have any init function that can add this, so for RPL list we need the static way of adding the data pointer.

File: subsys/emds/Kconfig [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+config EMDS_SECTOR_COUNT
+	int "Sector count of the emergency data storage area"
+	default 1
+	help
+	  Number of sectors used for the emergency data storage area
+
+choice
+	prompt "Initial log level limit"
+	default EMDS_LOG_LEVEL_WRN
```
Comment (tejlmand) [inline subsys/emds/Kconfig:31]: why are you not using the logging template here ?

For example like this:
https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/bluetooth/Kconfig.scan#L158-L160

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Pull request closed without merging**